### PR TITLE
Adding ReadOnlySession

### LINKED ISSUE #178: Add support for read only transactions
The neo4j java driver provides a method on AsyncSession to allow a user to specify that a transaction should not mutate data on the neo4j server: [ReadTransactionAsync](https://neo4j.com/docs/driver-manual/current/session-api/asynchronous/). This allows optimisation in clustered environments - such as routing read-only queries to read replicas.

It should be noted that introspection does not occur on the neo4j driver side, and it does not check to see that a query won't mutate data before routing the request, but the read replica will reject it.

### PR DESCRIPTION
**Pull request by geoffjohn11**
Closes #178

### CODE REVIEW

**Comment by geoffjohn11** [comment]
Need to add tests

**Review by BalmungSan — APPROVED** [review]
:+1:

Now we need to discuss and decide how we will use this need `readOnlyTransact`.
I see three options:

1. Use the read-only transact under the hood of all the read operations like single and list, and if someone wants to write something during a list, it has to be done using `Session.transact`.
**Pros**: Clean API, less work and it could improve the performance of some users.
**Cons**: We may end up breaking other people code and force them to change the API they use.

2. Leave the code as it is, so if someone wants to use a read-only transaction it has to be done explicitly using `Session.readOnlyTransact`.
**Pros**: We can be sure we are not breaking anything and code using read-only transactions is explicit about it.
**Cons**: Assuming most people are already doing read-only operations when using things like `single` and `list`, they would need to change the code to take advantage of this.

3. Add variants like `readOnlySingle`.
**Pros**: We can be sure we are not breaking anything and code using read-only transactions is explicit about it. Also, it may be said that this option provides a more standard API.
**Cons**: The same as the previous one, plus more code to maintain.

@geoffjohn11 as the author of the change, which one do you like the most?
@dimafeng what do you think?
@bengraygh you were the one that wanted this feature, what are your thoughts on this?
I will be honest that I would like option 1, IMHO one shouldn't be doing writing inside a reading operation, so for me, it makes sense to make that use case more explicit.

-----

BTW, another important discussion about this change is who will integrate this with #164 
I believe #164 is already big enough so I would prefer if this waits until we merge that one and then in this PR do the integration, but not sure what @geoffjohn11 thinks about that.

**Comment by geoffjohn11** [comment]
> @geoffjohn11 as the author of the change, which one do you like the most?

I think option 1 is my preference.

> BTW, another important discussion about this change is who will integrate this with #164
> I believe #164 is already big enough so I would prefer if this waits until we merge that one and then in this PR do the integration, but not sure what @geoffjohn11 thinks about that.

Sure, this change can wait.

**Comment by dimafeng** [comment]
I like option 1 too.

**Comment by bengraygh** [comment]
Bit late to the party, but I agree that option 1 is the right choice here.

**Comment by BalmungSan** [comment]
@geoffjohn11 hope you didn't mind me hijacking this PR I was reviewing your last change and I was about to propose that you should just use `guarantee`
But then I realized the previous implementation has a problem; except for **Futures** that are eager, the inner `F` you created inside `work`  will never be executed; I just fixed that as well as some failing tests.

After #164 is merged this PR will have a fun time solving conflicts and implementing `readOnlyTransact` for the **StreamingSession** 😬

File: core/src/main/scala/neotypes/Session.scala [hunk 1]
```diff
@@ -44,7 +50,30 @@ object Session {
...
+      lock.acquire.flatMap { _ =>
+        F.async[F[T]] { cb =>
+          val work: AsyncTransactionWork[CompletionStage[Unit]] = (tx: NeoAsyncTransaction) => {
+            val cf = new CompletableFuture[Unit]()
+
+            cb(Right(
+              F.delay(Transaction(F, tx)(lock)).guarantee(txf) {
+                case (tx, None)    => tx.commit.flatMap(_ => F.delay(internal.utils.void(cf.complete(()))))
+                case (tx, Some(_)) => tx.rollback.flatMap(_ => F.delay(internal.utils.void(cf.complete(()))))
+              }
```
Comment (BalmungSan) [inline core/src/main/scala/neotypes/Session.scala:66]: #164 will introduce a helper function called `txFinalizer` which we should use here like `.guarantee(txf)(txFinalizer.andThen(_.flatMap(_ => F.delay(internal.utils.void(cf.complete(()))))))`

**Comment by geoffjohn11** [comment]
> But then I realized the previous implementation has a problem; except for **Futures** that are eager, the inner `F` you created inside `work` will never be executed; I just fixed that as well as some failing tests.

Apologies for this mistake @BalmungSan !  I took a look cats-effect and the implementation of guarantee is calls `.use(f)` which says:  Allocates a resource and supplies it to the given function.

This is the reason why `guarantee` evaluates the F.delay effect?

**Comment by BalmungSan** [comment]
@geoffjohn11 

> Apologies for this mistake @BalmungSan

No worries, I reviewed the change like three times and never noticed it. Mixing pure and impure code is hard!

> This is the reason why guarantee evaluates the F.delay effect?

Not really, if you look what I made was an `Async` that returned an **F** and then I flattened it.
This is one of the wonders of FP, computation is a value so I just returned it into the effect system to execute it in the appropriate moment.

**Comment by BalmungSan** [comment]
Sadly also closed due #221 because of https://github.com/neo4j/neo4j-java-driver/issues/797

**Pull request closed without merging**